packages/service-worker/config/src/generator.ts

```diff
@@ -45,36 +45,44 @@ export class Generator {
 
   private async processAssetGroups(config: Config, hashTable: {[file: string]: string|undefined}):
       Promise<Object[]> {
+    // Retrieve all files of the build.
+    const allFiles = await this.fs.list('/');
     const seenMap = new Set<string>();
-    return Promise.all((config.assetGroups || []).map(async (group) => {
+    const filesPerGroup = new Map<AssetGroup, string[]>();
+
+    // Computed which files belong to each asset-group.
+    for (const group of (config.assetGroups || [])) {
       if ((group.resources as any).versionedFiles) {
         throw new Error(
             `Asset-group '${group.name}' in 'ngsw-config.json' uses the 'versionedFiles' option, ` +
             'which is no longer supported. Use \'files\' instead.');
       }
 
       const fileMatcher = globListToMatcher(group.resources.files || []);
-      const allFiles = await this.fs.list('/');
-
       const matchedFiles = allFiles.filter(fileMatcher).filter(file => !seenMap.has(file)).sort();
+
       matchedFiles.forEach(file => seenMap.add(file));
+      filesPerGroup.set(group, matchedFiles);
+    }
 
-      // Add the hashes.
-      await matchedFiles.reduce(async (previous, file) => {
-        await previous;
-        const hash = await this.fs.hash(file);
-        hashTable[joinUrls(this.baseHref, file)] = hash;
-      }, Promise.resolve());
+    // Compute hashes for all matched files and add them to the hash-table.
+    const allMatchedFiles = ([] as string[]).concat(...Array.from(filesPerGroup.values())).sort();
```

Comment:
I think the concat is unnecessary here since the `values()` method will return a new array each time:

```suggestion
    const allMatchedFiles = Array.from(filesPerGroup.values()).sort();
```

Reply:
`Array.from(filesPerGroup.values())` is an array of arrays. The `.concat()` is for flattening it :grin:

Reply:
Ahah! I didn't notice the `...`. Fair enough